File: solutions/415.add-strings/src/solution.ts
/*
 * @lc app=leetcode.cn id=415 lang=typescript
 *
 * https://leetcode.cn/problems/add-strings/
 */

// @lc code=start
function addStrings(num1: string, num2: string): string {
    let nums = [];
    let carry = 0;
    let i = num1.length - 1;
    let j = num2.length - 1;
    while (i >= 0 || j >= 0) {
        const n1 = i >= 0 ? parseInt(num1[i]) : 0;
        const n2 = j >= 0 ? parseInt(num2[j]) : 0;
        const sum = n1 + n2 + carry;
        carry = sum >= 10 ? 1 : 0;
        nums.push(sum % 10);
        i--;
        j--;
    }
    if (carry) {
        nums.push(carry);
    }
    return nums.reverse().join('');
}
// @lc code=end

export { addStrings };

File: solutions/415.add-strings/src/solution.rs
/*
 * @lc app=leetcode.cn id=415 lang=rust
 *
 * https://leetcode.cn/problems/add-strings/
 */

pub struct Solution;

// @lc code=start
impl Solution {
    pub fn add_strings(num1: String, num2: String) -> String {
        let num_1: Vec<char> = num1.chars().collect();
        let num_2: Vec<char> = num2.chars().collect();
        let mut nums = vec![];
        let mut carry = 0;
        let mut i = num1.len() as i32 - 1;
        let mut j = num2.len() as i32 - 1;
        while i >= 0 || j >= 0 {
            let n_1 = if i >= 0 {
                num_1[i as usize].to_digit(10).unwrap_or(0)
            } else {
                0
            };
            let n_2 = if j >= 0 {
                num_2[j as usize].to_digit(10).unwrap_or(0)
            } else {
                0
            };
            let sum = n_1 + n_2 + carry;
            carry = sum / 10;
            nums.push(std::char::from_digit(sum % 10, 10).unwrap());
            i -= 1;
            j -= 1;
        }
        if carry > 0 {
            nums.push(std::char::from_digit(carry, 10).unwrap());
        }
        nums.reverse();
        return nums.into_iter().collect();
    }
}
// @lc code=end
